crates/subspace-farmer/src/dsn.rs

```diff
@@ -43,52 +39,91 @@ impl DSNSync for NoSync {
     }
 }
 
-// TODO: Can we know in advance amount of all pieces in the network? That would simplify the whole
-// algorithm as it would simply request all pieces within expected range of pieces for which we can
-// get rewards from the consensus.
 /// Syncs pieces from the provided DSN.
 ///
-/// It would query pieces going upwards and downwards. It will adjust the requested piece range in
-/// order to ask the same amount of data over the network.
+/// It would query pieces going upwards and downwards from the address provided in options.
 pub async fn sync<DSN, OP>(mut dsn: DSN, options: SyncOptions, on_pieces: OP) -> anyhow::Result<()>
 where
     DSN: DSNSync + Send + Sized,
     DSN::Stream: Unpin + Send,
     // On pieces might `break` with some result from the sync or just continue
     OP: FnMut(FlatPieces, Vec<PieceIndex>) -> ControlFlow<anyhow::Result<()>> + Send + 'static,
 {
-    let mut range_size = options.initial_range_size;
-    let mut increasing_cursor = PieceIndexHashNumber::from(Sha256Hash::from(options.address));
+    let SyncOptions {
+        range_size,
+        address,
+    } = options;
+    let mut increasing_cursor = PieceIndexHashNumber::from(Sha256Hash::from(address));
     let mut decreasing_cursor = increasing_cursor;
     let mut increasing = true;
     let on_pieces = Arc::new(Mutex::new(on_pieces));
 
     loop {
```

Comment:
Looking at this it is somewhat hard to comprehend. Can we split input range into two non-overflowing segments and go through them separately without necessity of having 4 branches. Also if API returns us cursor, there would be no need for having `range_size`, you'd just continue from latest returned piece until sync returns you empty set or piece outside of range that you care about.

Reply:
You will still need to have those 4 branches in the end. I just don't see how having a cursor would benefit here.

As for splitting input range into 2 non-overflowing segments, it would be harder to follow, I think. Having `yield` and generators might benefit readability here.

Reply:
Hm... for me it is harder to follow two cursors at the same time with `increasing` boolean flipping around. It just doesn't read cleanly start to finish.

With separated non-overlapping ranges there would be two loops that sync each of segments independently, one at a time, then inside there will be an `if` statement with a termination check.

As an exercise try to explain the current workflow with plain English text. BTW, it would be handy to have it as a comment here. Often more complex workflow is harder to explain in the comment. By "explain" I mean not to re-iterate the code, but rather what it tries to achieve.

Reply:
Another benefit of the cursor is that you can get a proper batch of transactions back. Otherwise you don't know how full the range you're requesting: might be very sparse, might be very dense.

Would be great for you to sync with Shamil on this one.

Reply:
Yeah, cursor is the implementation of the DSN. I just mean that this isn't the interface :full_moon_with_face:, so it shouldn't be on the trait

Reply:
If it is an interface of the DSN, then `get_pieces` should return something that contains a cursor in it, right?

Either way, at least add explanation section for the algorithm here.

Reply:
Srry, updated docs now